trinity/sync/full/trie.py

```diff
@@ -0,0 +1,231 @@
+        if node_key in self._existing_nodes:
+            self.logger.debug("Node %s already exists in db" % encode_hex(node_key))
+            return
+        if await self.db.coro_exists(node_key):
+            self._existing_nodes.add(node_key)
+            self.logger.debug("Node %s already exists in db" % encode_hex(node_key))
+            return
+        self._schedule(node_key, parent, depth, leaf_callback, is_raw)
+
+    def _schedule(self, node_key: Hash32, parent: SyncRequest, depth: int,
+                  leaf_callback: Callable[[bytes, 'SyncRequest'], Awaitable[None]],
+                  is_raw: bool = False) -> None:
+        if parent is not None:
+            parent.dependencies += 1
+
+        existing = self.requests.get(node_key)
+        if existing is not None:
+            self.logger.debug(
+                "Already requesting %s, will just update parents list" % node_key)
+            existing.parents.append(parent)
+            return
+
+        request = SyncRequest(node_key, parent, depth, leaf_callback, is_raw)
+        # Requests get added to both self.queue and self.requests; the former is used to keep
+        # track which requests should be sent next, and the latter is used to avoid scheduling a
+        # request for a given node multiple times.
+        self.logger.debug("Scheduling retrieval of %s" % encode_hex(request.node_key))
+        self.requests[request.node_key] = request
+        bisect.insort(self.queue, request)
+
+    def get_children(self, request: SyncRequest
+                     ) -> Tuple[List[Tuple[int, Hash32]], List[bytes]]:
+        """Return all children of the node retrieved by the given request.
+
+        :rtype: A two-tuple with one list containing the children that reference other nodes and
+        another containing the leaf children.
+        """
+        node = decode_node(request.data)
+        return _get_children(node, request.depth)
+
+    async def process(self, results: List[Tuple[Hash32, bytes]]) -> None:
+        """Process request results.
+
+        :param results: A list of two-tuples containing the node's key and data.
+        """
+        for node_key, data in results:
+            request = self.requests.get(node_key)
+            if request is None:
+                # This may happen if we resend a request for a node after waiting too long,
+                # and then eventually get two responses with it.
+                self.logger.debug(
+                    "No SyncRequest found for %s, maybe we got more than one response for it"
+                    % encode_hex(node_key))
+                return
+
+            if request.data is not None:
+                raise SyncRequestAlreadyProcessed("%s has been processed already" % request)
+
+            request.data = data
+            if request.is_raw:
+                await self.commit(request)
+                continue
+
+            references, leaves = self.get_children(request)
```

Comment:
Related to my other comment about `request.data` being potentially problematic.  At this call-site we'd instead pass in a `response` object so that we have a guarantee that the `data` attribute is present.